AIP-84 Get Configs / Get Config Value

jason810496 opened the pull request:
closes: #42745
related: #42370

jason810496 committed:
AIP-84 Get Config

jason810496 committed:
Fix section param type and text format

jason810496 committed:
Add test for Get Config

jscheffl commented on airflow/api_fastapi/core_api/routes/public/config.py:
```diff
@@ -0,0 +1,122 @@
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN,
+            detail="Your Airflow administrator chose not to expose the configuration, most likely for security reasons.",
+        )
```
These lines are redundant to lines 46-62 - can you push this to a common function and use this from both methods?

jason810496 replied:
Sure, just resolved. Thanks for the reminder!

jason810496 committed:
Consolidate redundant for check expose config

pierrejeambrun commented on airflow/api_fastapi/core_api/routes/public/config.py:
```diff
@@ -0,0 +1,121 @@
+)
+def get_config(
+    section: str | None = None,
+    content_type: Literal["application/json", "text/plain"] = Header(...),
```
Nice.

Mimetype on fastapi are already defined in the `dag_source.py` file. What we need to do is extract from `dag_source.py` and this file the `mimetype literal` or enum, and put that in a common file, most likely `common/types.py`.

As you will see we also need the wildcard mimetype `*/*` because some http client send that to mention that they accept everything. In that case we should most likely default to `application/json`.

And then re-use that literal in both here and `dag_sources`.

Lastly you can use the `Annotated` notation, which is more modern and `Annotated[MimeTypeLiteral, Header()]`. (Not sure if the ellipsis `...` is needed).

And remove the `NOT_ACCEPTABLE` code path for the `dag_source`. (handled by pydantic validation now)

pierrejeambrun replied:
Also as you will see in `dag_source`. We use `start_with` because sometimes the mimetype can be `application/json; charset=utf-8` and not an exact match with `application_json`, but i didn't look to much to it.

I think some tests were failing otherwise.

jason810496 replied:
Thanks for pointing that out ! 
I’ll refactor the header query into a common module.

pierrejeambrun commented on airflow/api_fastapi/core_api/routes/public/config.py:
```diff
@@ -0,0 +1,121 @@
+    "/",
+    responses=create_openapi_http_exception_doc(
+        [status.HTTP_401_UNAUTHORIZED, status.HTTP_403_FORBIDDEN, status.HTTP_404_NOT_FOUND]
+    ),
```
You should add example responses because there is no way for FastAPI to populate correctly the openapi spec and swagger based on that.

You can take a look at how we do it in `get_dag_source`. (We combine `responses` for text/plain and `response_model` for application/json documentation)

pierrejeambrun commented on airflow/api_fastapi/core_api/routes/public/config.py:
```diff
@@ -0,0 +1,121 @@
+    "/section/{section}/option/{option}",
+    responses=create_openapi_http_exception_doc(
+        [status.HTTP_401_UNAUTHORIZED, status.HTTP_403_FORBIDDEN, status.HTTP_404_NOT_FOUND]
+    ),
```
Same for docs.

pierrejeambrun commented on tests/api_fastapi/core_api/routes/public/test_config.py:
```diff
@@ -0,0 +1,317 @@
+            (SECTION_NOT_EXIST, HEADERS_JSON, 404, {"detail": f"Section {SECTION_NOT_EXIST} not found."}),
+        ],
+    )
+    def test_get_config(self, test_client, section, headers, expected_status_code, expected_response):
+        query_params = {"section": section} if section else None
+        if expected_status_code == 403:
+            with conf_vars(AIRFLOW_CONFIG_DISABLE_EXPOSE_CONFIG):
+                response = test_client.get("/public/config/", headers=headers, params=query_params)
+        else:
+            response = test_client.get("/public/config/", headers=headers, params=query_params)
```
I think we are missing a test with `non-sensitive-only` ?

jason810496 replied:
It seems the legacy API missed the `non-sensitive-only` case in the `get_config` endpoint. I’ll add this case along with a test. Thanks !

pierrejeambrun replied:
I think you can take inspiration from `test_should_respond_200_text_plain_with_non_sensitive_only` from legacy tests.

pierrejeambrun reviewed (commented):
Nice.

A few suggestions but looking good.

jason810496 committed:
Add `Accept` header depends for Json and Text

jason810496 committed:
Refactor config, dag_source with common header

jason810496 committed:
Refactor test_config with non-sensitive-only case

pierrejeambrun commented on airflow/api_fastapi/common/headers.py:
```diff
@@ -0,0 +1,48 @@
+        Header(
+            json_schema_extra={
+                "type": "string",
+                "enum": [Mimetype.JSON, Mimetype.TEXT],
```
Are we missing the `Mimetype.Any` ? (maybe dynamically retrieve the `Mimetype` Enum values), or maybe it supports an enum out of the box ?

pierrejeambrun commented on airflow/api_fastapi/core_api/openapi/v1-generated.yaml:
```diff
@@ -1439,6 +1442,154 @@ paths:
+
+                  base_log_folder = /opt/airflow/logs
+
+                  '
```
Here we need at least 2 sections (as in the legacy doc) to show that multiples sections are returned in comparison to the other endpoint.

pierrejeambrun commented on airflow/api_fastapi/core_api/openapi/v1-generated.yaml:
```diff
@@ -1439,6 +1442,154 @@ paths:
+              schema:
+                type: string
+                example: '[core]
+
```
Why do we have line break between every lines in the example ?

pierrejeambrun replied:
Also should we only have one value here ?

jason810496 replied:
Hi @pierrejeambrun 
The OpenAPI schema at `http://localhost:29091/docs#/Config/get_config` appears to be formatted correctly:  
<img width="544" alt="get_config OpenAPI schema" src="https://github.com/user-attachments/assets/866ebd28-1385-4017-ae3a-4fd6652d861d">


The line breaks are introduced by the auto-generation script (`scripts/ci/pre_commit/update_fastapi_api_spec.py`, which runs `scripts/in_container/run_update_fastapi_api_spec.py` using `get_openapi` provided by `FastAPI`).  

I prefer not to manually remove the line breaks because once this PR is merged, the line breaks will be reintroduced when others run the pre-commit script. wdyt ?

pierrejeambrun replied:
Got it, no worry then, as long as the doc appears properly

pierrejeambrun commented on tests/api_fastapi/core_api/routes/public/test_config.py:
```diff
@@ -0,0 +1,460 @@
+
+from tests_common.test_utils.config import conf_vars
+
+HEADERS_JSON = {"Accept": "application/json"}
+HEADERS_TEXT = {"Accept": "text/plain"}
+HEADERS_INVALID = {"Accept": "invalid"}
+HEADERS_JSON_UTF8 = {"Accept": "application/json; charset=utf-8"}
```
We are missing a test with `*/*`  I think

pierrejeambrun commented on tests/api_fastapi/core_api/routes/public/test_dag_sources.py:
```diff
@@ -79,7 +79,9 @@ def test_should_respond_200_text(self, test_client):
 
-    @pytest.mark.parametrize("headers", [{"Accept": "application/json"}, {}])
+    @pytest.mark.parametrize(
+        "headers", [{"Accept": "application/json"}, {"Accept": "application/json; charset=utf-8"}, {}]
```
Here too `*/*` should behave as `application/json`

pierrejeambrun reviewed (commented):
Thanks looking good, a few suggestions / question, then we can merge.

jason810496 committed:
Fix OpenAPI schema for Headers, Get Config

jason810496 committed:
Refactor test_config

- add `HEADERS_NONE` and `HEADERS_ANY` cases
- modularize common json response cases
- add `_validate_response` common method

pierrejeambrun reviewed (approved):
Good to merge after rebase

rawwar commented on tests/api_fastapi/core_api/routes/public/test_dag_sources.py:
```diff
@@ -74,11 +74,14 @@ def test_should_respond_200_text(self, test_client, test_dag):
+    @pytest.mark.parametrize(
+        "headers", [{"Accept": "application/json"}, {"Accept": "application/json; charset=utf-8"}, {}]
+    )
+    def test_should_respond_200_json(self, test_client,test_dag, headers):
```
```suggestion
    def test_should_respond_200_json(self, test_client, test_dag, headers):
```

rawwar commented on airflow/api_fastapi/core_api/routes/public/dag_sources.py:
```diff
@@ -18,21 +18,20 @@
 
-from fastapi import Depends, Header, HTTPException, Request, Response, status
+from fastapi import Depends, HTTPException, Request, Response, status
+from itsdangerous import BadSignature, URLSafeSerializer
```
```suggestion
from fastapi import Depends, HTTPException, Response, status
```

rawwar replied:
I ran pre-commits locally and ruff gave this fix

jason810496 replied:
I think it cause by my manual resolve conflict when rebasing to latest main. Just resolve, thanks !

jason810496 committed:
Fix style by ruff format
